# **Shark Attacks Data Cleaning Exercise - The Bridge Data Science Bootcamp 2021-2022**

The objective of the exercise is to clean the data from Global Sharks Attacks in Kaggle:  https://www.kaggle.com/teajay/global-shark-attacks with the following conditions:

 + No column can be deleted
 + The resulting csv can´t have less than 1,500 row but it is preferable if we can keep around 25% of total rows (~ 6,400 rows)
 + No nans are allowed
 + Each of the columns shall contain the same data type

**Index**

- **1. Data overview and overall cleaning** 


- **2. Colums data cleaning**
    * 2.1 Columns 22 and 23
    * 2.2 Date related columns: Case Number(s), Date, Year and pdf
    * 2.3 Species
    * 2.4 Sex and Name
    * 2.5 Activity
    * 2.6 Age
    * 2.7 Columns Injury and Fatal
    * 2.8 Columns Location, Area and Country 
    * 2.9 Type
    * 2.10 Colums href and href formula
    * 2.11 Investigator or Source
    * 2.12 Time
    
    
- **3. Reordering, correction and saving**

## 1. Data Overview and overall cleaning

In [1]:
import pandas as pd
import numpy as np
import re
import warnings

warnings.filterwarnings("ignore")
pd.set_option('display.max_columns', None)  # to show all columns in the Data Frame


In [2]:
df=pd.read_csv('attacks.csv', encoding='cp1252') # encoding is not always necessary, just try it if it doesn't work without it

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25723 entries, 0 to 25722
Data columns (total 24 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Case Number             8702 non-null   object 
 1   Date                    6302 non-null   object 
 2   Year                    6300 non-null   float64
 3   Type                    6298 non-null   object 
 4   Country                 6252 non-null   object 
 5   Area                    5847 non-null   object 
 6   Location                5762 non-null   object 
 7   Activity                5758 non-null   object 
 8   Name                    6092 non-null   object 
 9   Sex                     5737 non-null   object 
 10  Age                     3471 non-null   object 
 11  Injury                  6274 non-null   object 
 12  Fatal (Y/N)             5763 non-null   object 
 13  Time                    2948 non-null   object 
 14  Species                 3464 non-null 

* From 25.723 rows the column with the maximum entries is column 0: "Case Number" -  with 8,702
* We have two colums with practically no data in them (22 and 23)

In [4]:
df

Case Number         Date    Year        Type    Country  \
0      2018.06.25  25-Jun-2018  2018.0     Boating        USA   
1      2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA   
2      2018.06.09  09-Jun-2018  2018.0     Invalid        USA   
3      2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA   
4      2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO   
...           ...          ...     ...         ...        ...   
25718         NaN          NaN     NaN         NaN        NaN   
25719         NaN          NaN     NaN         NaN        NaN   
25720         NaN          NaN     NaN         NaN        NaN   
25721         NaN          NaN     NaN         NaN        NaN   
25722          xx          NaN     NaN         NaN        NaN   

                  Area                        Location     Activity  \
0           California     Oceanside, San Diego County     Paddling   
1              Georgia  St. Simon Island, Glynn County     Standing   
2               Hawaii                    Habush, Oahu      Surfing   
3      New South Wales              Arrawarra Headland      Surfing   
4               Colima                        La Ticla  Free diving   
...                ...                             ...          ...   
25718              NaN                             NaN          NaN   
25719              NaN                             NaN          NaN   
25720              NaN                             NaN          NaN   
25721              NaN                             NaN          NaN   
25722              NaN                             NaN          NaN   

                  Name Sex   Age  \
0          Julie Wolfe    F   57   
1      Adyson McNeely     F   11   
2          John Denges    M   48   
3                 male    M  NaN   
4       Gustavo Ramos     M  NaN   
...                ...  ...  ...   
25718              NaN  NaN  NaN   
25719              NaN  NaN  NaN   
25720              NaN  NaN  NaN   
25721              NaN  NaN  NaN   
25722              NaN  NaN  NaN   

                                                  Injury Fatal (Y/N)  \
0      No injury to occupant, outrigger canoe and pad...           N   
1                             Minor injury to left thigh           N   
2           Injury to left lower leg from surfboard skeg           N   
3                              Minor injury to lower leg           N   
4      Lacerations to leg & hand shark PROVOKED INCIDENT           N   
...                                                  ...         ...   
25718                                                NaN         NaN   
25719                                                NaN         NaN   
25720                                                NaN         NaN   
25721                                                NaN         NaN   
25722                                                NaN         NaN   

                Time         Species           Investigator or Source  \
0              18h00      White shark                R. Collier, GSAF   
1      14h00  -15h00              NaN  K.McMurray, TrackingSharks.com   
2              07h45              NaN  K.McMurray, TrackingSharks.com   
3                NaN        2 m shark                  B. Myatt, GSAF   
4                NaN  Tiger shark, 3m                       A .Kipper   
...              ...              ...                             ...   
25718            NaN              NaN                             NaN   
25719            NaN              NaN                             NaN   
25720            NaN              NaN                             NaN   
25721            NaN              NaN                             NaN   
25722            NaN              NaN                             NaN   

                            pdf  \
0          2018.06.25-Wolfe.pdf   
1        2018.06.18-McNeely.pdf   
2         2018.06.09-Denges.pdf   
3      2018.06.08-Arrawarra.pdf   
4          2018.06.04-

* Last rows of the data frame seem not to contain any data

* First of all we will delete the rows for which all values are null

In [5]:
df = df.dropna(how='all')
df

Case Number         Date    Year        Type    Country  \
0      2018.06.25  25-Jun-2018  2018.0     Boating        USA   
1      2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA   
2      2018.06.09  09-Jun-2018  2018.0     Invalid        USA   
3      2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA   
4      2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO   
...           ...          ...     ...         ...        ...   
8698            0          NaN     NaN         NaN        NaN   
8699            0          NaN     NaN         NaN        NaN   
8700            0          NaN     NaN         NaN        NaN   
8701            0          NaN     NaN         NaN        NaN   
25722          xx          NaN     NaN         NaN        NaN   

                  Area                        Location     Activity  \
0           California     Oceanside, San Diego County     Paddling   
1              Georgia  St. Simon Island, Glynn County     Standing   
2               Hawaii                    Habush, Oahu      Surfing   
3      New South Wales              Arrawarra Headland      Surfing   
4               Colima                        La Ticla  Free diving   
...                ...                             ...          ...   
8698               NaN                             NaN          NaN   
8699               NaN                             NaN          NaN   
8700               NaN                             NaN          NaN   
8701               NaN                             NaN          NaN   
25722              NaN                             NaN          NaN   

                  Name Sex   Age  \
0          Julie Wolfe    F   57   
1      Adyson McNeely     F   11   
2          John Denges    M   48   
3                 male    M  NaN   
4       Gustavo Ramos     M  NaN   
...                ...  ...  ...   
8698               NaN  NaN  NaN   
8699               NaN  NaN  NaN   
8700               NaN  NaN  NaN   
8701               NaN  NaN  NaN   
25722              NaN  NaN  NaN   

                                                  Injury Fatal (Y/N)  \
0      No injury to occupant, outrigger canoe and pad...           N   
1                             Minor injury to left thigh           N   
2           Injury to left lower leg from surfboard skeg           N   
3                              Minor injury to lower leg           N   
4      Lacerations to leg & hand shark PROVOKED INCIDENT           N   
...                                                  ...         ...   
8698                                                 NaN         NaN   
8699                                                 NaN         NaN   
8700                                                 NaN         NaN   
8701                                                 NaN         NaN   
25722                                                NaN         NaN   

                Time         Species           Investigator or Source  \
0              18h00      White shark                R. Collier, GSAF   
1      14h00  -15h00              NaN  K.McMurray, TrackingSharks.com   
2              07h45              NaN  K.McMurray, TrackingSharks.com   
3                NaN        2 m shark                  B. Myatt, GSAF   
4                NaN  Tiger shark, 3m                       A .Kipper   
...              ...              ...                             ...   
8698             NaN              NaN                             NaN   
8699             NaN              NaN                             NaN   
8700             NaN              NaN                             NaN   
8701             NaN              NaN                             NaN   
25722            NaN              NaN                             NaN   

                            pdf  \
0          2018.06.25-Wolfe.pdf   
1        2018.06.18-McNeely.pdf   
2         2018.06.09-Denges.pdf   
3      2018.06.08-Arrawarra.pdf   
4          2018.06.04-

* 8703 resulting rows
* Still, at the bottom there are some rows with all values nules but for the column "Case Number" with either 0s or "xx". Lets proceed to see all values in this column 

In [6]:
pd.set_option('display.max_rows',10) 
df['Case Number'].value_counts()

0               2400
1920.00.00.b       2
1966.12.26         2
2014.08.02         2
1990.05.10         2
                ... 
1999.09.05         1
1999.09.10         1
1999.09.16         1
1999.09.18         1
xx                 1
Name: Case Number, Length: 6287, dtype: int64

* There are 2400 '0's and one 'xx' value which we will substitute for Nans and then delete all rows with all Nans again

In [7]:
df['Case Number'] = df['Case Number'].replace(["0", 'xx' ], np.nan)

In [8]:
df = df.dropna(how='all')

In [9]:
df

Case Number         Date    Year        Type    Country             Area  \
0     2018.06.25  25-Jun-2018  2018.0     Boating        USA       California   
1     2018.06.18  18-Jun-2018  2018.0  Unprovoked        USA          Georgia   
2     2018.06.09  09-Jun-2018  2018.0     Invalid        USA           Hawaii   
3     2018.06.08  08-Jun-2018  2018.0  Unprovoked  AUSTRALIA  New South Wales   
4     2018.06.04  04-Jun-2018  2018.0    Provoked     MEXICO           Colima   
...          ...          ...     ...         ...        ...              ...   
6304         NaN          NaN     NaN         NaN        NaN              NaN   
6305         NaN          NaN     NaN         NaN        NaN              NaN   
6306         NaN          NaN     NaN         NaN        NaN              NaN   
6307         NaN          NaN     NaN         NaN        NaN              NaN   
6308         NaN          NaN     NaN         NaN        NaN              NaN   

                            Location     Activity             Name Sex   Age  \
0        Oceanside, San Diego County     Paddling      Julie Wolfe    F   57   
1     St. Simon Island, Glynn County     Standing  Adyson McNeely     F   11   
2                       Habush, Oahu      Surfing      John Denges    M   48   
3                 Arrawarra Headland      Surfing             male    M  NaN   
4                           La Ticla  Free diving   Gustavo Ramos     M  NaN   
...                              ...          ...              ...  ...  ...   
6304                             NaN          NaN              NaN  NaN  NaN   
6305                             NaN          NaN              NaN  NaN  NaN   
6306                             NaN          NaN              NaN  NaN  NaN   
6307                             NaN          NaN              NaN  NaN  NaN   
6308                             NaN          NaN              NaN  NaN  NaN   

                                                 Injury Fatal (Y/N)  \
0     No injury to occupant, outrigger canoe and pad...           N   
1                            Minor injury to left thigh           N   
2          Injury to left lower leg from surfboard skeg           N   
3                             Minor injury to lower leg           N   
4     Lacerations to leg & hand shark PROVOKED INCIDENT           N   
...                                                 ...         ...   
6304                                                NaN         NaN   
6305                                                NaN         NaN   
6306                                                NaN         NaN   
6307                                                NaN         NaN   
6308                                                NaN         NaN   

               Time         Species           Investigator or Source  \
0             18h00      White shark                R. Collier, GSAF   
1     14h00  -15h00              NaN  K.McMurray, TrackingSharks.com   
2             07h45              NaN  K.McMurray, TrackingSharks.com   
3               NaN        2 m shark                  B. Myatt, GSAF   
4               NaN  Tiger shark, 3m                       A .Kipper   
...             ...              ...                             ...   
6304            NaN              NaN                             NaN   
6305            NaN              NaN                             NaN   
6306            NaN              NaN                             NaN   
6307            NaN              NaN                             NaN   
6308            NaN              NaN                             NaN   

                           pdf  \
0         2018.06.25-Wolfe.pdf   
1       2018.06.18-McNeely.pdf   
2        2018.06.09-Denges.pdf   
3     2018.06.08-Arrawarra.pdf   
4         2018.06.04-Ramos.pdf   
...                        ...   
6304                       NaN   
6305                       NaN   
6306                       NaN   
6307         

* Still, at the bottom of the results there are some rows with only the original order number (from 6304 to 6310).
* Lets delete these rows ase they are not usefull at all.

In [10]:
df = df[df["original order"] < 6304]
df

Case Number         Date    Year        Type             Country  \
0     2018.06.25  25-Jun-2018  2018.0     Boating                 USA   
1     2018.06.18  18-Jun-2018  2018.0  Unprovoked                 USA   
2     2018.06.09  09-Jun-2018  2018.0     Invalid                 USA   
3     2018.06.08  08-Jun-2018  2018.0  Unprovoked           AUSTRALIA   
4     2018.06.04  04-Jun-2018  2018.0    Provoked              MEXICO   
...          ...          ...     ...         ...                 ...   
6297     ND.0005  Before 1903     0.0  Unprovoked           AUSTRALIA   
6298     ND.0004  Before 1903     0.0  Unprovoked           AUSTRALIA   
6299     ND.0003    1900-1905     0.0  Unprovoked                 USA   
6300     ND.0002    1883-1889     0.0  Unprovoked              PANAMA   
6301     ND.0001    1845-1853     0.0  Unprovoked  CEYLON (SRI LANKA)   

                   Area                             Location      Activity  \
0            California          Oceanside, San Diego County      Paddling   
1               Georgia       St. Simon Island, Glynn County      Standing   
2                Hawaii                         Habush, Oahu       Surfing   
3       New South Wales                   Arrawarra Headland       Surfing   
4                Colima                             La Ticla   Free diving   
...                 ...                                  ...           ...   
6297  Western Australia                          Roebuck Bay        Diving   
6298  Western Australia                                  NaN  Pearl diving   
6299     North Carolina                       Ocracoke Inlet      Swimming   
6300                NaN                 Panama Bay 8ºN, 79ºW           NaN   
6301   Eastern Province  Below the English fort, Trincomalee      Swimming   

                       Name Sex   Age  \
0               Julie Wolfe    F   57   
1           Adyson McNeely     F   11   
2               John Denges    M   48   
3                      male    M  NaN   
4            Gustavo Ramos     M  NaN   
...                     ...  ...  ...   
6297                   male    M  NaN   
6298                  Ahmun    M  NaN   
6299  Coast Guard personnel    M  NaN   
6300        Jules Patterson    M  NaN   
6301                   male    M   15   

                                                 Injury Fatal (Y/N)  \
0     No injury to occupant, outrigger canoe and pad...           N   
1                            Minor injury to left thigh           N   
2          Injury to left lower leg from surfboard skeg           N   
3                             Minor injury to lower leg           N   
4     Lacerations to leg & hand shark PROVOKED INCIDENT           N   
...                                                 ...         ...   
6297                                              FATAL           Y   
6298                                              FATAL           Y   
6299                                              FATAL           Y   
6300                                              FATAL           Y   
6301  FATAL. "Shark bit him in half, carrying away t...           Y   

               Time         Species                  Investigator or Source  \
0             18h00      White shark                       R. Collier, GSAF   
1     14h00  -15h00              NaN         K.McMurray, TrackingSharks.com   
2             07h45              NaN         K.McMurray, TrackingSharks.com   
3               NaN        2 m shark                         B. Myatt, GSAF   
4               NaN  Tiger shark, 3m                              A .Kipper   
...             ...              ...                                    ...   
6297            NaN              NaN       H. Taunton; N. Bartlett,  p. 234   
6298            NaN              NaN  H. Taunton; N. Bartlett,  pp. 233-234   
6299            NaN              NaN   F. Schwartz, p.23; C. Creswell, GSAF   
6300            NaN              NaN                   

## 2.Colums Data Cleaning

* Now that all rows seems to have at least some usefull information we will proceed to clean the data by columns or related clusters of columns. 

## 2.1 Columns 22 and 23

* First we will clean the most "empty" columns: Unnamed 22 and Unnamed 23

* Lets see what do we have in the valid values for this columns

In [11]:
df[~df['Unnamed: 22']. isna()]

Case Number         Date    Year        Type Country    Area  \
1478  2006.05.27  27-May-2006  2006.0  Unprovoked     USA  Hawaii   

                Location Activity          Name Sex  Age  \
1478  North Shore, O'ahu  Surfing  Bret Desmond    M  31   

                                 Injury Fatal (Y/N)   Time Species   \
1478  No injury, shark bumped surfboard           N  16h00      NaN   

     Investigator or Source                     pdf  \
1478             R. Collier  2006.05.27-Desmond.pdf   

                                           href formula  \
1478  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                   href Case Number.1  \
1478  http://sharkattackfile.net/spreadsheets/pdf_di...    2006.05.27   

     Case Number.2  original order   Unnamed: 22 Unnamed: 23  
1478    2006.05.27          4825.0  stopped here         NaN

* We see that the only alid value of column 22 is the string "stopped here"

* Seems a reference and not relevent for the dataframe, however lets verify if this value repeats within the dataframe

In [12]:
df[(df == 'stopped here').any(axis=1)]

Case Number         Date    Year        Type Country    Area  \
1478  2006.05.27  27-May-2006  2006.0  Unprovoked     USA  Hawaii   

                Location Activity          Name Sex  Age  \
1478  North Shore, O'ahu  Surfing  Bret Desmond    M  31   

                                 Injury Fatal (Y/N)   Time Species   \
1478  No injury, shark bumped surfboard           N  16h00      NaN   

     Investigator or Source                     pdf  \
1478             R. Collier  2006.05.27-Desmond.pdf   

                                           href formula  \
1478  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                   href Case Number.1  \
1478  http://sharkattackfile.net/spreadsheets/pdf_di...    2006.05.27   

     Case Number.2  original order   Unnamed: 22 Unnamed: 23  
1478    2006.05.27          4825.0  stopped here         NaN

* There is not such a value in all the Data Frame, so we will delete it

* First let see whats in column 23 and repeat the process that we followed in column 22

In [13]:
df[~df['Unnamed: 23']. isna()]

Case Number                  Date    Year        Type  \
4415    1952.03.30           30-Mar-1952  1952.0  Unprovoked   
5840  1878.09.14.R  Reported 14-Sep-1878  1878.0    Provoked   

                   Country         Area                    Location  \
4415  NETHERLANDS ANTILLES      Curacao                         NaN   
5840                   USA  Connecticut  Branford, New Haven County   

                                             Activity              Name Sex   \
4415  Went to aid of child being menaced by the shark       A.J. Eggink    M   
5840                                          Fishing  Captain Pattison    M   

      Age                                        Injury Fatal (Y/N) Time  \
4415  NaN                Buttock bitten, tissue removed           N  NaN   
5840  NaN  Leg bitten by netted shark PROVOKED INCIDENT           N  NaN   

                                               Species   \
4415  Bull shark, 2.7 m [9'] was captured & dragged ...   
5840                                                NaN   

                                 Investigator or Source  \
4415  J. Randall, p.352 in Sharks & Survival; H.D. B...   
5840                       St. Joseph Herald, 9/14/1878   

                            pdf  \
4415      1952.03.30-Eggink.pdf   
5840  1878.09.14.R-Pattison.pdf   

                                           href formula  \
4415  http://sharkattackfile.net/spreadsheets/pdf_di...   
5840  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                   href Case Number.1  \
4415  http://sharkattackfile.net/spreadsheets/pdf_di...    1952.03.30   
5840  http://sharkattackfile.net/spreadsheets/pdf_di...  1878.09.14.R   

     Case Number.2  original order Unnamed: 22      Unnamed: 23  
4415    1952.03.30          1888.0         NaN           Teramo  
5840  1878.09.14.R           463.0         NaN  change filename

* We find the strings "Teramo" and "change filename".
* Change filename seems a reference, lets check if Teramo repeats within the dataframe and if we can delete it

In [14]:
df[(df == 'Teramo').any(axis=1)]

Case Number         Date    Year        Type               Country  \
4403  1952.08.05  05-Aug-1952  1952.0    Provoked                 ITALY   
4415  1952.03.30  30-Mar-1952  1952.0  Unprovoked  NETHERLANDS ANTILLES   

         Area    Location                                         Activity  \
4403   Teramo  Giulianova                                          Fishing   
4415  Curacao         NaN  Went to aid of child being menaced by the shark   

                Name Sex   Age                               Injury  \
4403  Vittorio Speca  NaN   19  Multiple injuries PROVOKED INCIDENT   
4415     A.J. Eggink    M  NaN       Buttock bitten, tissue removed   

     Fatal (Y/N)   Time                                           Species   \
4403           Y  02h00                                           2m shark   
4415           N    NaN  Bull shark, 2.7 m [9'] was captured & dragged ...   

                                 Investigator or Source  \
4403                                     C. Moore, GSAF   
4415  J. Randall, p.352 in Sharks & Survival; H.D. B...   

                        pdf  \
4403   1952.08.05-Speca.pdf   
4415  1952.03.30-Eggink.pdf   

                                           href formula  \
4403  http://sharkattackfile.net/spreadsheets/pdf_di...   
4415  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                   href Case Number.1  \
4403  http://sharkattackfile.net/spreadsheets/pdf_di...    1952.08.04   
4415  http://sharkattackfile.net/spreadsheets/pdf_di...    1952.03.30   

     Case Number.2  original order Unnamed: 22 Unnamed: 23  
4403    1952.08.05          1900.0         NaN         NaN  
4415    1952.03.30          1888.0         NaN      Teramo

* There is another row with 'Teramo' (row 4403)

* As Teramo is a region in Italy, where attack in row 4403 occurs and both attacks occur in the same year we will assume that this is a human introduction error and delete the value "Teramo"

**Normally we would delete columns 22 and 23 as they do not offer any info, but as the exercise indicates not to delete any column and also that we cannot have NaNs then we will input the string "No Data" in both Unnamed22 and Unnamed23 colums so that they dont make any impact in the data_frame** 

In [15]:
df['Unnamed: 22'] = "No Data"
df['Unnamed: 23'] = "No Data"

## 2.2 Date related columns: Case Number(s), Date, Year and pdf

* If we check these columns we realize that all contain info on the date, so lets try to complete the data in all of them avoiding missmatches between columns

In [16]:
df[["Case Number", "Case Number.1", "Case Number.2", "Date", "Year" , "pdf"]]

Case Number Case Number.1 Case Number.2         Date    Year  \
0     2018.06.25    2018.06.25    2018.06.25  25-Jun-2018  2018.0   
1     2018.06.18    2018.06.18    2018.06.18  18-Jun-2018  2018.0   
2     2018.06.09    2018.06.09    2018.06.09  09-Jun-2018  2018.0   
3     2018.06.08    2018.06.08    2018.06.08  08-Jun-2018  2018.0   
4     2018.06.04    2018.06.04    2018.06.04  04-Jun-2018  2018.0   
...          ...           ...           ...          ...     ...   
6297     ND.0005       ND.0005       ND.0005  Before 1903     0.0   
6298     ND.0004       ND.0004       ND.0004  Before 1903     0.0   
6299     ND.0003       ND.0003       ND.0003    1900-1905     0.0   
6300     ND.0002       ND.0002       ND.0002    1883-1889     0.0   
6301     ND.0001       ND.0001       ND.0001    1845-1853     0.0   

                                 pdf  
0               2018.06.25-Wolfe.pdf  
1             2018.06.18-McNeely.pdf  
2              2018.06.09-Denges.pdf  
3           2018.06.08-Arrawarra.pdf  
4               2018.06.04-Ramos.pdf  
...                              ...  
6297          ND-0005-RoebuckBay.pdf  
6298               ND-0004-Ahmun.pdf  
6299  ND-0003-Ocracoke_1900-1905.pdf  
6300      ND-0002-JulesPatterson.pdf  
6301              ND-0001-Ceylon.pdf  

[6302 rows x 6 columns]

* We see that colums Case Number, Case Number 1 and case number 3 are almost identical
* Also we see that there is a pattern as the pdf column is a combination of: Case Number + "-" + Name + ".pdf"

* First lets check if Case Number(s) colums are identical

In [17]:
df['Case Number.1'].equals(df['Case Number.2'].equals(df['Case Number']))

False

In [18]:
df['Case Number.1'].isin(df['Case Number.2']).value_counts()

True     6283
False      19
Name: Case Number.1, dtype: int64

In [19]:
df['Case Number.1'].isin(df['Case Number']).value_counts()

True     6279
False      23
Name: Case Number.1, dtype: int64

In [20]:
df['Case Number.2'].isin(df['Case Number']).value_counts()

True     6298
False       4
Name: Case Number.2, dtype: int64

* Case number(s) columns do not contain identical info. 
* Case number 2 and Case Number are the most similar with only 4 different values, lets start unnifying these first

In [21]:
df_compare = df[~df['Case Number.2'].isin(df['Case Number'])] # creates a new dataframe for comparing purpuses
df_compare[['Case Number', 'Case Number.2', 'Case Number.1', 'Date', 'Year', 'pdf']]

Case Number Case Number.2 Case Number.1                  Date    Year  \
390     2015.07-10    2015.07.10    2015.07.10           10-Jul-2015  2015.0   
4949  1934.01.08.R  1934.02.08.R  1934.02.08.R  Reported 08-Feb-1934  1934.0   
5488           NaN  1905.09.06.R  1905.09.06.R  Reported 06-Sep-1905  1905.0   
5944       1864.05    1864.05.00    1864.05.00              May-1864  1864.0   

                           pdf  
390      2015.07.10-Miskin.pdf  
4949   1924.02.08.R-Turkey.pdf  
5488    1905.09.06.R-Curry.pdf  
5944  1864.05.00.Mauritius.pdf

* We see that for these values Case Number seem to contain errors if we compare to the other values, so we will unify imputing the values of Case Number.2 into Case Number

In [22]:
df['Case Number'] = df ['Case Number.2']

* Now lets see diference between unified Case Number and Case Number 1

In [23]:
df['Case Number.2'].isin(df['Case Number.1']).value_counts()

True     6282
False      20
Name: Case Number.2, dtype: int64

In [24]:
df_compare = df[~df['Case Number.2'].isin(df['Case Number.1'])]
df_compare[['Case Number', 'Case Number.2', 'Case Number.1', 'Date', 'Year', 'pdf']]

Case Number Case Number.2 Case Number.1                   Date    Year  \
34      2018.04.03    2018.04.03    2018.04.02            03-Apr-2018  2018.0   
117   2017.07.20.a  2017.07.20.a  2017/07.20.a            20-Jul-2017  2017.0   
144     2017.05.06    2017.05.06    2017.06.06            05-May-2017  2017.0   
217     2016.09.15    2016.09.15    2016.09.16            16-Sep-2016  2016.0   
314   2016.01.24.b  2016.01.24.b  2015.01.24.b            24-Jan-2016  2016.0   
...            ...           ...           ...                    ...     ...   
5386  1911.07.31.R  1911.07.31.R  1911.07.31.T   Reported 31-Jul-1911  1911.0   
5777  1884.04.38.R  1884.04.38.R  1884.04.28.R   Reported 28-Apr-1884  1884.0   
5883  1876.06.18.R  1876.06.18.R  1874.06.18.R  Reported 18-Jun- 1874  1874.0   
6009    1853,94.29    1853,94.29    1853.04.29            29-Apr-1853  1853.0   
6122  1808.06.01.R  1808.06.01.R  1808.05.01.R   Reported 01-May-1808  1808.0   

                               pdf  
34     2018.04.03-StFrancisBay.pdf  
117       2017.07.20.a-Geyling.pdf  
144          2017.05.06-Rozada.pdf  
217      2016.09.16-BellsBeach.pdf  
314          2016.01.24.b-Love.pdf  
...                            ...  
5386        1911.07.31.R-Ceuta.pdf  
5777    1884.04.28.R-ChalaCook.pdf  
5883  1874.06.18.R-DiazServant.pdf  
6009         1853.04.29-Hanson.pdf  
6122      1808.05.01.R-Antigua.pdf  

[20 rows x 6 columns]

* In this case we observe that for some attacks the colum "Case Number.1" is wrong (see above index 314) if we compare with the the values in columns  "Year", "Date" and "pdf",  and for other attacks the column "Case Number" is incorrect (see index 5883 above).

* For these values we will take for valid the case number in the "pdf" column which seems to coincide with "Date" and "Year" columns

* Therefore we set up a formula to rename all missmatches between Case Number.2 and Case.Number.1 with the relevant part of the pdf column values 

In [25]:
df_compare['Case Number'] = df_compare['pdf'].str.split('-') # First we substitute "Case Number" with the values from "pdf" in list format 

In [26]:
df_compare['Case Number'] 

34       [2018.04.03, StFrancisBay.pdf]
117         [2017.07.20.a, Geyling.pdf]
144            [2017.05.06, Rozada.pdf]
217        [2016.09.16, BellsBeach.pdf]
314            [2016.01.24.b, Love.pdf]
                     ...               
5386          [1911.07.31.R, Ceuta.pdf]
5777      [1884.04.28.R, ChalaCook.pdf]
5883    [1874.06.18.R, DiazServant.pdf]
6009           [1853.04.29, Hanson.pdf]
6122        [1808.05.01.R, Antigua.pdf]
Name: Case Number, Length: 20, dtype: object

In [27]:
new_values=[]
for i in df_compare['Case Number']:
    new_values.append(i[0])

In [28]:
df_compare.drop("Case Number", axis = 1, inplace = True)

df_compare["Case Number"] = new_values
df_compare[["Case Number", "pdf"]]

Case Number                           pdf
34      2018.04.03   2018.04.03-StFrancisBay.pdf
117   2017.07.20.a      2017.07.20.a-Geyling.pdf
144     2017.05.06         2017.05.06-Rozada.pdf
217     2016.09.16     2016.09.16-BellsBeach.pdf
314   2016.01.24.b         2016.01.24.b-Love.pdf
...            ...                           ...
5386  1911.07.31.R        1911.07.31.R-Ceuta.pdf
5777  1884.04.28.R    1884.04.28.R-ChalaCook.pdf
5883  1874.06.18.R  1874.06.18.R-DiazServant.pdf
6009    1853.04.29         1853.04.29-Hanson.pdf
6122  1808.05.01.R      1808.05.01.R-Antigua.pdf

[20 rows x 2 columns]

* Now that we have all Case Numbers missmatches inserted properly in df_compare we will transfer them to df (our main data frame) and equal "Case Number" columns 1 and 2 to the main "Case Number" column 

In [29]:
df.loc[df["pdf"].isin(df_compare["pdf"]), ['Case Number']] = df_compare[['Case Number']] # substitute in main df from df_compare only if pdf column is equal

In [30]:
df['Case Number'] = df['Case Number'].str.replace('-','.') #to unify naming, substitute all "-" with "."
df["Case Number.1"]=df["Case Number"]
df["Case Number.2"]=df["Case Number"]

* We have the columns Case Number cleaned, but before moving on lets check if we have any duplicates

In [31]:
df[df['Case Number'].isin(df['Case Number'][df['Case Number'].duplicated()])]

Case Number                  Date    Year        Type Country  \
522       2014.08.02           02-Aug-2014  2014.0  Unprovoked     USA   
523       2014.08.02           02-Aug-2014  2014.0  Unprovoked     USA   
615       2013.10.05           06-Oct-2013  2013.0  Unprovoked     USA   
616       2013.10.05           10-Oct-2013  2013.0  Unprovoked     USA   
746     2012.09.02.b           02-Sep-2012  2012.0  Unprovoked     USA   
...              ...                   ...     ...         ...     ...   
5323  1915.07.06.a.R  Reported 06-Jul-1915  1915.0  Unprovoked  MEXICO   
5354    1913.08.27.R  Reported 27-Aug-1913  1913.0     Invalid     USA   
5355    1913.08.27.R  Reported 27-Aug-1913  1913.0     Invalid     USA   
5443    1907.10.16.R  Reported 16-Oct-1907  1907.0  Unprovoked   CHINA   
5444    1907.10.16.R  Reported 16-Oct-1907  1907.0  Unprovoked   CHINA   

            Area                                         Location  \
522      Florida             South of Cocoa Beach, Brevard County   
523      Florida                      Table Beach, Brevard County   
615   California   Bunkers, Humboldt Bay, Eureka, Humboldt County   
616      Florida                          Destin, Okaloosa County   
746      Florida                 New Smyrna Beach, Volusia County   
...          ...                                              ...   
5323         NaN                                  Santa Maria Bar   
5354  New Jersey                          Lavalette, Ocean County   
5355  New Jersey                     Spring Lake, Monmouth County   
5443   Hong Kong  Sharp Peak, Sai Kung Peninsula, New Territories   
5444   Hong Kong  Sharp Peak, Sai Kung Peninsula, New Territories   

             Activity                      Name Sex   Age  \
522           Surfing                      male    M  50s   
523   Boogie boarding        Christian Sanhueza    M    8   
615           Surfing              Jay Scrivner    M   45   
616            Wading  Zachary Tyke Standridge     M   12   
746   Boogie boarding                    female    F    8   
...               ...                       ...  ...  ...   
5323           Wading             J.W. McDonald    M  NaN   
5354              NaN                       NaN    M  NaN   
5355              NaN                       NaN    F  NaN   
5443          Fishing                 fishermen    M  NaN   
5444          Fishing                 fishermen    M  NaN   

                                                 Injury Fatal (Y/N)   Time  \
522                                         Foot bitten           N    NaN   
523                                 Laceration to ankle           N  13h00   
615                                 Laceration to thigh           N  08h45   
616                       Lacerations to  right forearm           N  15h30   
746                    Puncture wounds to calf and hand           N  18h30   
...                                                 ...         ...    ...   
5323                                              FATAL           Y    NaN   
5354              Man's leg recovered from 800-lb shark         NaN    NaN   
5355                   Female foot recovered from shark         NaN    NaN   
5443  3 of thel 5 were injured, one of whom lost bot...           N    NaN   
5444   2 of the 5 fishermen were so seriously injure...           Y    NaN   

                                          Species   \
522                                            NaN   
523                                            NaN   
615                         White shark, 8' to 10'   
616                               Small bull shark   
746                               3.5' to 4' shark   
...                                            ...   
5323                                           NaN   
5354  Shark involvement prior to death unconfirmed   
5355  Shark involvement prior to death unconfirmed   
5443                    Shark involvement probable   
5444                    Sha

The repeated cases numbers are not identical and correspond to different individuals attacked in the same attack, so we leave it as it is 


 - **Now we have homogeneous "Cases Number" naming:**
     - aaaa.mm.dd for normal dates (plus another classifiers if there were in the pdf name)
     - for attacks with no certain date: ND.xxxx 
     - for attacks Before Christ: 0000.aaaa

## Column Date


In [32]:
pd.set_option('display.max_rows', None)  # to show all rows in the Data Frame
df['Date'].value_counts()

1957                                                                11
1942                                                                 9
1956                                                                 8
1958                                                                 7
1950                                                                 7
1941                                                                 7
1949                                                                 6
No date                                                              6
Oct-1960                                                             5
12-Apr-2001                                                          5
1970s                                                                5
28-Jul-1995                                                          5
No date, Before 1963                                                 5
1940                                                                 5
05-Oct

* We can see that we don´t have only exact dates, but also approximations for the cases that the date is not certain


* As we cannot delete any column and we have already the date in the Case Numbers columns we will:
    1. rename "Date" column to "Date Details" in order to keep the additional info on the dates
    2. Use "Case Number 2" column to have only the info on the date with no other references. For this we will:
        * Rename "Case Number 2" column to "Date"
        * Remove any other identification characters from the new "Date" column 

In [33]:
df=df.rename(columns={"Date": "Date Details"})
df=df.rename(columns={"Case Number.2": "Date"})
df['Date'].value_counts() # to check status of Date column (previous Case Number 2)

1920.00.00.b          2
1915.07.06.a.R        2
2009.12.18            2
2014.08.02            2
1962.06.11.b          2
1990.05.10            2
2013.10.05            2
1907.10.16.R          2
1913.08.27.R          2
2005.04.06            2
1980.07.00            2
1983.06.15            2
1966.12.26            2
1923.00.00.a          2
2006.09.02            2
2012.09.02.b          2
1957.02.05            1
1957.04.13            1
1957.09.02            1
1957.08.00            1
1957.04.07.a          1
1957.06.21            1
1957.07.24            1
1957.04.07.b          1
1957.07.09            1
1957.07.15            1
1957.04.22            1
1957.05.11            1
1957.04.23            1
1957.04.28            1
1957.05.00            1
1957.05.07            1
1957.02.24            1
2018.06.25            1
1957.02.02            1
1957.02.00            1
1956.10.07            1
1956.10.20            1
1956.10.27            1
1956.12.09            1
1956.12.15.R          1
1956.12.24      

Now lets assign to Case Numbers ND.xxxx the value: "No date" and to delete any alphanumerical characters in the rest of the values.

In [34]:
#Assign to entries "ND.xxxx" the value: "No date" and for rest of attacks delete all characters with index > 9 (or len 10)
def transform_date(date): 
    counter_date=0
    for i in date:
        if i.startswith("ND"):
            df['Date'][counter_date]= "No Date"
            counter_date+=1
        else:
            if len(i)>10:
                date[counter_date]= df['Date'][counter_date][:10]
                counter_date+=1
            else:
                counter_date+=1

transform_date(df['Date'])

df['Date']

0       2018.06.25
1       2018.06.18
2       2018.06.09
3       2018.06.08
4       2018.06.04
5       2018.06.03
6       2018.06.03
7       2018.05.27
8       2018.05.26
9       2018.05.26
10      2018.05.24
11      2018.05.21
12      2018.05.13
13      2018.05.13
14      2018.05.00
15      2018.05.12
16      2018.05.09
17      2018.04.30
18      2018.04.28
19      2018.04.28
20      2018.04.25
21      2018.04.25
22      2018.04.24
23      2018.04.23
24      2018.04.22
25      2018.04.19
26      2018.04.15
27      2018.04.15
28      2018.04.15
29      2018.04.15
30      2018.04.14
31      2018.04.10
32      2018.04.09
33      2018.04.05
34      2018.04.03
35      2018.03.31
36      2018.03.14
37      2018.03.09
38      2018.03.09
39      2018.02.24
40      2018.02.23
41      2018.02.17
42      2018.02.15
43      2018.02.14
44      2018.02.11
45      2018.02.03
46      2018.02.01
47      2018.01.28
48      2018.01.21
49      2018.01.14
50      2018.01.13
51      2018.01.12
52      2018

Now we have "clean" dates where possible

## Column Year

In [35]:
df['Year']

0       2018.0
1       2018.0
2       2018.0
3       2018.0
4       2018.0
5       2018.0
6       2018.0
7       2018.0
8       2018.0
9       2018.0
10      2018.0
11      2018.0
12      2018.0
13      2018.0
14      2018.0
15      2018.0
16      2018.0
17      2018.0
18      2018.0
19      2018.0
20      2018.0
21      2018.0
22      2018.0
23      2018.0
24      2018.0
25      2018.0
26      2018.0
27      2018.0
28      2018.0
29      2018.0
30      2018.0
31      2018.0
32      2018.0
33      2018.0
34      2018.0
35      2018.0
36      2018.0
37      2018.0
38      2018.0
39      2018.0
40      2018.0
41      2018.0
42      2018.0
43      2018.0
44      2018.0
45      2018.0
46      2018.0
47      2018.0
48      2018.0
49      2018.0
50      2018.0
51      2018.0
52      2018.0
53      2017.0
54      2017.0
55      2017.0
56      2017.0
57      2017.0
58      2017.0
59      2017.0
60      2017.0
61      2017.0
62      2017.0
63      2017.0
64      2017.0
65      2017.0
66      20

* Column year seems to present the following issues:
     * We still have Nans
     * Numbers are float, not int
     * Some years Before Christ and "No Date" values are represented by 0s
 
* First lets fix the NaNs: 

In [36]:
df[df['Year']. isna()]

Case Number          Date Details  Year        Type    Country  \
187   2017.01.08.R  Reported 08-Jan-2017   NaN     Invalid  AUSTRALIA   
6079  1836.08.19.R  Reported 19-Aug-1836   NaN  Unprovoked    ENGLAND   

            Area    Location      Activity          Name Sex   Age  \
187   Queensland         NaN  Spearfishing  Kerry Daniel    M   35   
6079  Cumberland  Whitehaven      Swimming         a boy    M  NaN   

                                      Injury Fatal (Y/N) Time    Species   \
187   No attack, shark made a threat display         NaN  NaN  Bull shark   
6079                                   FATAL           Y  NaN         NaN   

      Investigator or Source                           pdf  \
187   Liquid Vision 1/8/2017  2017.01.08.R-KerryDaniel.pdf   
6079          C. Moore, GSAF   1835.08.19.R-Whitehaven.pdf   

                                           href formula  \
187   http://sharkattackfile.net/spreadsheets/pdf_di...   
6079  http://sharkattackfile.net/spreadsheets/pdf_di...   

                                                   href Case Number.1  \
187   http://sharkattackfile.net/spreadsheets/pdf_di...  2017.01.08.R   
6079  http://sharkattackfile.net/spreadsheets/pdf_di...  1836.08.19.R   

            Date  original order Unnamed: 22 Unnamed: 23  
187   2017.01.08          6116.0     No Data     No Data  
6079  1836.08.19           224.0     No Data     No Data

* We have 2 Nans in the "Year" column that can be changed manually as we have it in the "Date" column
* **However, to verify that we don´t have any data missmatches among date related columns we will extract the year from our Date column, also we will input BC values as negative years**. For this we will create a new list to extract the year values from 'Date', transform them and put them into the 'Year' column

In [37]:
year_list = df['Date'].to_list()

# Function to transform the Date column values that are into year_list into objective Year values: (see below)
def transform_year(years):
    counter_year=0
    for i in year_list:
        if i.startswith("No Date"): #if Date=="No Date" --> Year = "No Year" 
            year_list[counter_year]= "No Year"
            counter_year+=1
        elif i.startswith("0000"): #if Date is a BC value (which is coded in date as 0000.aaaa) --> 
            # --> Year = -aaaa (ignoring first 5 characters)  
            year_list[counter_year]= -int(year_list[counter_year][5:])
            counter_year+=1
        else:                      # otherwise it takes the year from "Date" column (first 4 characters)
            year_list[counter_year]= int(year_list[counter_year][:4])
            counter_year+=1
  
            
transform_year(year_list)

year_list

df['Year'] = year_list # we pass the results from the list to the "Year" column

df['Year']


0          2018
1          2018
2          2018
3          2018
4          2018
5          2018
6          2018
7          2018
8          2018
9          2018
10         2018
11         2018
12         2018
13         2018
14         2018
15         2018
16         2018
17         2018
18         2018
19         2018
20         2018
21         2018
22         2018
23         2018
24         2018
25         2018
26         2018
27         2018
28         2018
29         2018
30         2018
31         2018
32         2018
33         2018
34         2018
35         2018
36         2018
37         2018
38         2018
39         2018
40         2018
41         2018
42         2018
43         2018
44         2018
45         2018
46         2018
47         2018
48         2018
49         2018
50         2018
51         2018
52         2018
53         2017
54         2017
55         2017
56         2017
57         2017
58         2017
59         2017
60         2017
61         2017
62      

* Now we have the Years correctly input, **however we can improve the Year column by replacing "No Year" values with approximations based on the "Date Details" column:**  

    * If in this column it figures one year we will input the year: ie: Before 1911 --> 1911
    * If two or more years appear we will input the average ie:'No date: After August 1926 and before 1936' --> 1931
    * If a decade appears ie. 1920s we will input the first year of the decade --> 1920
    * If the word war appears we will input 1943
    * If any other character appears near a year we will input the year excluding the character: ie 1920? --> 1920

In [38]:
date_details = [] #Create a list to extract Date Details column ONLY for attacks with No Date and work with it
counter=0
for x in df['Date Details']:
    if df['Date'][counter] == "No Date":
        date_details.append(re.split('/ | |-',x))
        counter+=1
        
    else:
        counter+=1
        
date_details # we see now the Date Details for "No Date" values in a list but splitted by every element


[['Before', '1939'],
 ['1990', 'or', '1991'],
 ['Before', '2016'],
 ['Before', 'Oct', '2009'],
 ['Before', '1934'],
 ['Before', '1934'],
 ['2009?'],
 ['Before', '1930'],
 ['1880', '1899'],
 ['Before', '1909'],
 ['Before', '2012'],
 ['Before', '1916'],
 ['Between', '', '', '1951', '1963'],
 ['Before', '1908'],
 ['Before', '1900'],
 ['Before', '1876'],
 ['Before', '2012'],
 ['Before', '2011'],
 ['Before', '2011'],
 ['Before', '2009'],
 ['Beforer', '1994'],
 ['Before', '1963'],
 ['1896', '1913'],
 ['Before', '1936'],
 ['Before', '08', 'Jun', '1912'],
 ['Before', '2012'],
 ['Before', '1911'],
 ['Before', '1901'],
 ['No', 'date,', 'late', '1960s'],
 ['Before', '2006'],
 ['Before', '2003'],
 ['Before', '2004'],
 ['Before', '1962'],
 ['1950s'],
 ['No', 'date,', 'Before', '1963'],
 ['2003?'],
 ['No', 'date'],
 ['No', 'date'],
 ['Before', 'Feb', '1998'],
 ['No', 'date,', 'Before', 'May', '1996'],
 ['No', 'date,', 'Before', 'Mar', '1995'],
 ['Before', '1996'],
 ['No', 'date,', 'Before', 'Aug', '

In [39]:
counter=0
date_details2=[0] * len(date_details) # as we want to keep the list of lists for comparison purpose we cerate a new list to work with it

for sublist in date_details:          # to iterate over the sublists in date_details list   
    lista=[]
    for val in sublist:               # this loop put all digits with len > 2 (the years) in  a separate list
        if val.isdigit():
            if len(val) > 2:
                lista.append(val)
                
    for i in range(0, len(lista)):    # transfrom the numbers in the newly created list to integers
        lista[i] = int(lista[i])
    
    
    if len(lista)>0:                  # if the list contain any number return the average 
        date_details2[counter] = int(sum(lista)/len(lista))
        counter+=1
    
    else:                             # if the list do not contain any number we have to guess it from the strings
        
        combined = '\t'.join(sublist)       # join all strings in sublist before performing  a regex match
        pattern = re.compile("[0-9]{4}")    # pattern for looking for 4 continous digits in the string
        match_war = ["War","war"]           # match condition to look for the word War
        if any(x in combined for x in match_war): # if detects the substring war returns 1943
            date_details2[counter] = 1943
            counter+=1
                    
        else:                                    # other wise if there is a year combined with other characters it extract the year
            for x in sublist:
                if pattern.match(x):
                    date_details2[counter] = "".join(re.findall(pattern, x))
                    
                else:                           # if not it returns the value it had originally in date details column 
                    date_details2[counter] = date_details[counter]
            counter+=1
                    

In [40]:
compare = "\n".join("{} {}".format(x, y) for x, y in zip(date_details, date_details2)) 

In [41]:
print(compare) # Now we compare the output with the input

['Before', '1939'] 1939
['1990', 'or', '1991'] 1990
['Before', '2016'] 2016
['Before', 'Oct', '2009'] 2009
['Before', '1934'] 1934
['Before', '1934'] 1934
['2009?'] 2009
['Before', '1930'] 1930
['1880', '1899'] 1889
['Before', '1909'] 1909
['Before', '2012'] 2012
['Before', '1916'] 1916
['Between', '', '', '1951', '1963'] 1957
['Before', '1908'] 1908
['Before', '1900'] 1900
['Before', '1876'] 1876
['Before', '2012'] 2012
['Before', '2011'] 2011
['Before', '2011'] 2011
['Before', '2009'] 2009
['Beforer', '1994'] 1994
['Before', '1963'] 1963
['1896', '1913'] 1904
['Before', '1936'] 1936
['Before', '08', 'Jun', '1912'] 1912
['Before', '2012'] 2012
['Before', '1911'] 1911
['Before', '1901'] 1901
['No', 'date,', 'late', '1960s'] 1960
['Before', '2006'] 2006
['Before', '2003'] 2003
['Before', '2004'] 2004
['Before', '1962'] 1962
['1950s'] 1950
['No', 'date,', 'Before', '1963'] 1963
['2003?'] 2003
['No', 'date'] ['No', 'date']
['No', 'date'] ['No', 'date']
['Before', 'Feb', '1998'] 1998
['No'

* We have the column almost clean, but still we have a few No Date outputs.
* To adress this we have the following possibilities:
    * Input the median of the dataset
    * Input the median of the attacks for which we didn´t have a certain date
* However it seems that where we have the missing values the adjacent years are "kind of" sorted from higher to lower:



- ['2003?'] 2003
- ['No', 'date'] ['No', 'date']
- ['No', 'date'] ['No', 'date']
- ['Before', 'Feb', '1998'] 1998
- ['No', 'date,', 'Before', 'May', '1996'] 1996
- ['No', 'date,', 'Before', 'Mar', '1995'] 1995
- ['Before', '1996'] 1996
- ['No', 'date,', 'Before', 'Aug', '1989'] 1989
- ['No', 'date,', 'Before', 'Aug', '1987'] 1987


AND

- ['Early', '1930s'] 1930
- ['Before', '1927'] 1927
- ['Between', '1918', '&', '1939'] 1928
- ['No', 'date'] ['No', 'date']
- ['No', 'date'] ['No', 'date']
- ['No', 'date'] ['No', 'date']
- ['1920', '', '1923'] 1921
- ['Before', '1921'] 1921
- ['Before', '1911'] 1911
- ['Before', '1921'] 1921


Therefore we will input the value of the average between the inmediate preceding year with the inmediate following year in the dataset (if any of these years are also "No Date" then we move to the following year to make the calculations: 

In [42]:
counter=0
for x in date_details2: # If the value is still a list (so it is not yet a year) we make the calculatins explained above
    if type(x) == list:
        n=1
        while type(date_details2[counter+n]) == list:
            n+=1
        date_details2[counter] = int((int(date_details2[counter-1])+int(date_details2[counter+n]))/2)
    
    counter+=1

In [43]:
compare = "\n".join("{} {}".format(x, y) for x, y in zip(date_details, date_details2))

In [44]:
print(compare)

['Before', '1939'] 1939
['1990', 'or', '1991'] 1990
['Before', '2016'] 2016
['Before', 'Oct', '2009'] 2009
['Before', '1934'] 1934
['Before', '1934'] 1934
['2009?'] 2009
['Before', '1930'] 1930
['1880', '1899'] 1889
['Before', '1909'] 1909
['Before', '2012'] 2012
['Before', '1916'] 1916
['Between', '', '', '1951', '1963'] 1957
['Before', '1908'] 1908
['Before', '1900'] 1900
['Before', '1876'] 1876
['Before', '2012'] 2012
['Before', '2011'] 2011
['Before', '2011'] 2011
['Before', '2009'] 2009
['Beforer', '1994'] 1994
['Before', '1963'] 1963
['1896', '1913'] 1904
['Before', '1936'] 1936
['Before', '08', 'Jun', '1912'] 1912
['Before', '2012'] 2012
['Before', '1911'] 1911
['Before', '1901'] 1901
['No', 'date,', 'late', '1960s'] 1960
['Before', '2006'] 2006
['Before', '2003'] 2003
['Before', '2004'] 2004
['Before', '1962'] 1962
['1950s'] 1950
['No', 'date,', 'Before', '1963'] 1963
['2003?'] 2003
['No', 'date'] 2000
['No', 'date'] 1999
['Before', 'Feb', '1998'] 1998
['No', 'date,', 'Before',

**Now we have the "No Date" years cleaned, so we can transfer them again to the Year colum**

In [45]:
counter_df=0
counter_list=0
for x in df['Date']:
    if x == "No Date": 
        df['Year'][counter_df]= date_details2[counter_list]
        counter_list+=1
        
    counter_df+=1

In [46]:
df['Year']=df['Year'].astype(int)
df['Year']

0       2018
1       2018
2       2018
3       2018
4       2018
5       2018
6       2018
7       2018
8       2018
9       2018
10      2018
11      2018
12      2018
13      2018
14      2018
15      2018
16      2018
17      2018
18      2018
19      2018
20      2018
21      2018
22      2018
23      2018
24      2018
25      2018
26      2018
27      2018
28      2018
29      2018
30      2018
31      2018
32      2018
33      2018
34      2018
35      2018
36      2018
37      2018
38      2018
39      2018
40      2018
41      2018
42      2018
43      2018
44      2018
45      2018
46      2018
47      2018
48      2018
49      2018
50      2018
51      2018
52      2018
53      2017
54      2017
55      2017
56      2017
57      2017
58      2017
59      2017
60      2017
61      2017
62      2017
63      2017
64      2017
65      2017
66      2017
67      2017
68      2017
69      2017
70      2017
71      2017
72      2017
73      2017
74      2017
75      2017
76      2017

## 2.3 Column Species

1. Check the different values in the column

In [47]:
pd.set_option('display.max_rows', None)
df['Species '].value_counts()

White shark                                                                                                                                                                                             163
Shark involvement prior to death was not confirmed                                                                                                                                                      105
Invalid                                                                                                                                                                                                 102
Shark involvement not confirmed                                                                                                                                                                          88
Tiger shark                                                                                                                                                                             

* We see that the data is very atomized and in different formats, before cleaning it lets rename the column as it has a blank space and fill the Nans with the word "Unknown"

In [48]:
df=df.rename(columns={"Species ": "Species"})
df['Species'] = df['Species'].fillna('Unknown')

1. The Species column contains data that includes some details on the attack and also the shark length

2. For not loosing this data while having a more compact Species classification, we will divide Species column in two separate colums:
    2.1 Species Type: with a Keyword showing only the species (For this we will rename Unamed column 22)
    2.2 Species info: where we will keep all the data from the initial Species column 

In [49]:
df=df.rename(columns={"Species": "Species Info"})
df=df.rename(columns={"Unnamed: 22": "Species Type"})
df["Species Type"]=df["Species Info"] # For the moment we will input in species Type the info in Species Info to transform it later on
pd.set_option('display.max_rows', 10)
df[["Species Type", "Species Info"]]

Species Type     Species Info
0         White shark      White shark
1             Unknown          Unknown
2             Unknown          Unknown
3           2 m shark        2 m shark
4     Tiger shark, 3m  Tiger shark, 3m
...               ...              ...
6297          Unknown          Unknown
6298          Unknown          Unknown
6299          Unknown          Unknown
6300          Unknown          Unknown
6301          Unknown          Unknown

[6302 rows x 2 columns]

In [50]:
pd.set_option('display.max_rows', None)
df['Species Type'].value_counts()

Unknown                                                                                                                                                                                                 2839
White shark                                                                                                                                                                                              163
Shark involvement prior to death was not confirmed                                                                                                                                                       105
Invalid                                                                                                                                                                                                  102
Shark involvement not confirmed                                                                                                                                                     

Transformation of Species Type column:

1. Create a dictionary with equivalences between some existing values in the dataset (as keys) and more concentrated objective values for the dataframe (as values)
2. After concentrating the values we will rename the rest to: "Unknown with details"  (as despite not having a "Species Type" value more info is in the "Species Info" column)

In [58]:
# First we define a list of allowed values for the category
species_map={'involvement':'shark involvement not confirmed',
              'invalid':'unknown without details',
              'grey nurse shark':'grey nurse',
                'nurse shark':'nurse',
                'grey reef shark':'grey reef',
                'white shark':'white',
                'bull':'bull',
                'tiger shark':'tiger',
                'hammerhead':"hammerhead",
                'whaler':"whaler",
                'oceanic whitetip': "oceanic whitetip",
                'thresher':"thresher",
                'caribbean reef':"caribbean reef",
                'blue':"blue",
                'lemon':"lemon",
                'spinner':"spinner",
                'wobbegong':"wobbegong",
                'sandtiger':"sandtiger",
                'blacktip':"blacktip",
                'porbeagle':"porbeagle",
                'mako':"mako",
                'salmon':"salmon",
                'sandbar':"sandbar",
                'zambesi':"zambesi",
                'goblin':"goblin",
                'raggedtooth':"raggedtooth",
                'port jackson':"port jackson",
                '7-gill':"7-gill",
                'seven':"7-gill",
                'cow':"cow",
                'silky':"silky",
                'galapagos':"galapagos",
                'cookiecutter':"cookiecutter",
                'leucas':"leucas",
                'carpet':"carpet",
                'whale shark':"whale",
                'reef shark':"reef",
                'juvenile':"juvenile",
                'dog':"dog",
                'basking':"basking",
                'porbeagle':"porbeagle",
                'sand':"sand",
                'thresher':"thresher",
                'silky':"silky",
                'dusky':"dusky"
             }
df["Species Type"] = df["Species Type"].str.lower()

def species_clean(species_list):
    for k in species_map:
        if k in species_list:
            return species_map[k]
    return "Unknown with details" 

df["Species Type"] = df["Species Type"].apply(species_clean)

Unknown with details               3966
white                               633
shark involvement not confirmed     332
tiger                               276
bull                                187
unknown without details             103
blacktip                             74
whaler                               72
nurse                                57
mako                                 53
blue                                 52
wobbegong                            50
hammerhead                           48
spinner                              47
raggedtooth                          43
grey nurse                           40
lemon                                34
zambesi                              29
oceanic whitetip                     24
reef                                 23
7-gill                               20
sand                                 19
caribbean reef                       17
grey reef                            17
dusky                                11


In [59]:
pd.set_option('display.max_rows', 50)
df["Species Type"] = df["Species Type"].str.capitalize()
df['Species Type'].value_counts()

Unknown with details               3966
White                               633
Shark involvement not confirmed     332
Tiger                               276
Bull                                187
Unknown without details             103
Blacktip                             74
Whaler                               72
Nurse                                57
Mako                                 53
Blue                                 52
Wobbegong                            50
Hammerhead                           48
Spinner                              47
Raggedtooth                          43
Grey nurse                           40
Lemon                                34
Zambesi                              29
Oceanic whitetip                     24
Reef                                 23
7-gill                               20
Sand                                 19
Caribbean reef                       17
Grey reef                            17
Dusky                                11


**We have now a much more concentrated column, as for example:**
   * White shark attacks increased form 163 to 633
   * Tiger attacks increased from 73 to 276
   * Bull from 52 to 187 

### 2.4 COLUMNS SEX AND NAME


* We will analize these columns together as the Name column contains also info on the gender

In [ ]:
pd.set_option('display.max_rows', None)
df['Name'].value_counts()

In [ ]:
df['Sex '].value_counts()

1. We have to rename the Sex column as it contains an Space
2. We will fill Nans of both columns with "Unknown"
3. As column Name contains info on the Gender, we will complete the Sex column with the values from Name column 

In [ ]:
df=df.rename(columns={"Sex ": "Sex"})

In [ ]:
CAMBIAR!!!!!!!


df['Sex'] = df['Sex'].fillna('Unknown')
df['Name'] = df['Name'].fillna('Unknown')

df.loc[df['Name'].str.startswith('male'), 'Sex'] = "M"



df.loc[df['Name'].str.contains(' male ', case=False), 'Sex'] = "F"
df.loc[df['Name'].str.contains('female', case=False), 'Sex'] = "F"
df.loc[df['Name'].str.contains(' boy ', case=False), 'Sex'] = "M"
df.loc[df['Name'].str.contains('sailor', case=False), 'Sex'] = "M"
df.loc[df['Name'].str.contains('girl', case=False), 'Sex'] = "F"
df.loc[df['Name'].str.contains('fisherm', case=False), 'Sex'] = "M"
df.loc[df['Name'].str.contains('women', case=False), 'Sex'] = "F"
df.loc[df['Name'].str.contains('woman', case=False), 'Sex'] = "F"



df.loc[df['Sex']=='N', 'Sex'] = "M"
df.loc[df['Sex']=='M ', 'Sex'] = "M"
df.loc[df['Sex']=='lli', 'Sex'] = "Unknown"
df.loc[df['Sex']==".", 'Sex'] = "Unknown"



In [ ]:
df['Sex'].value_counts()

**We slightly increased our male registers from 5094 to 5100 and our female registers from 637 to 653**

## 2.5 Column Activity

In [ ]:
df['Activity'].value_counts().head(30)

* First we will Replace the null values with 'unknown' 
* Then we will concentrate the activities utilizing certain Keywords as the dataset is very atomized

In [ ]:
df['Activity'].fillna('unknown', inplace = True)
df['Activity'] = df['Activity'].str.lower().str.replace('-',' ') #as some activites were the same but separated by a "-" instead of a blank space
df['Activity'].value_counts().head(40)

In [ ]:
df.loc[df['Activity'].str.contains('swimming', case=False), 'Activity'] = 'swimming'
df.loc[df['Activity'].str.contains('board', case=False), 'Activity'] = 'surfing'
df.loc[df['Activity'].str.contains('surf', case=False), 'Activity'] = 'surfing'
df.loc[df['Activity'].str.contains('fishing', case=False), 'Activity'] = 'fishing'
df.loc[df['Activity'].str.contains('diving', case=False), 'Activity'] = 'diving'
df.loc[df['Activity'].str.contains('boat', case=False), 'Activity'] = 'boat'
df.loc[df['Activity'].str.contains('jump', case=False), 'Activity'] = 'jumping into the water'
df.loc[df['Activity'].str.contains('bathing', case=False), 'Activity'] = 'swimming'


df['Activity'].value_counts().head(30)


## 2.6 Column Age

In [ ]:
df['Age'].value_counts()

* For this column we see that for some values we have a string containing an age range or, if different people were attacked, the age for each of them.

* In order to unify the data we will create a list for each value of the dataset containing all elements in the initial string.


In [ ]:
df['Age'] = df['Age'].astype('string')

df['Age'] = df['Age'].str.split(' ')

df['Age']


* Then we will return the average of the values in each list if there were numbers in the list

* If the list doesn´t contain any number we will return NaN to work with this data in further steps

In [ ]:
def clean(z):
    lista=[]
    for x in z:
        if x.isdigit(): 
            lista.append(x)
    for i in range(0, len(lista)):
        lista[i] = int(lista[i])
    
    if len(lista)>0:
        return int(sum(lista)/len(lista)) 
    
    else:
        return np.NaN

df['Age'] = df['Age'].fillna('Unknown')
    
df['Age'] = df['Age'].apply(clean)

In [ ]:
df['Age']

Now to fill the NaNs we will take a look at the age distribution of:

1. the whole data set
2. The Data set by Sex
3. The Data set by Activity

In [ ]:
import matplotlib
df.Age.plot(kind='kde')

In [ ]:
data_aux = {}

df_aux = pd.DataFrame(data_aux)
df_aux['Age'] = df['Age']
df_aux['Activity'] = df['Activity']
df_aux['Sex'] = df['Sex']


threshold = 100 # Anything that occurs less than this will be removed.
for col in df.Activity:
    value_counts = df_aux.Activity.value_counts() # Specific column 
    to_remove = value_counts[value_counts <= threshold].index
    df_aux.Activity.replace(to_remove, np.nan, inplace=True)

df_aux.groupby('Sex').Age.plot(kind='kde')

In [ ]:
df_aux.groupby('Activity').Age.plot(kind='kde')

* We see that the age distribution changes depending on the Sex and the Activity, therefore we will implement the following layers for substituting the Nans:
    1. We will input the median age for each sex in each activity
    2. For the values that were not possible to complete (as we dont have cases of both sex´s in each activity) we will attend only to the activity, as it has a higher variability than the Sex
    3. For the rest of the cases we will attend only to the sex


In [ ]:
df['Age'] = df.groupby(['Activity', 'Sex'])['Age'].apply(lambda x: x.fillna(x.median()))
df['Age'] = df.groupby(['Activity'])['Age'].apply(lambda x: x.fillna(x.median()))
df['Age'] = df.groupby(['Sex'])['Age'].apply(lambda x: x.fillna(x.median()))
df.Age.plot(kind='kde')

## 2.7 Columns Injury and Fatal

* We wil analize these columns together as the column Injury contains usefull info for completing the column fatal

In [ ]:
df['Injury'].value_counts()

In [ ]:
df['Fatal (Y/N)'].value_counts()

* Both columns are already quite clean, we will just:
    * Complete column "FATAL Y/N" for all the cases in which the injury was fatal.
    * fix the incorrect values Y/N (where possible) and the Nans replace with "unknown". 

In [ ]:
df['Fatal (Y/N)']=df['Fatal (Y/N)'].replace(' N', 'N')
df['Fatal (Y/N)']=df['Fatal (Y/N)'].replace('M', 'N')
df['Fatal (Y/N)']=df['Fatal (Y/N)'].replace('N ', 'N')
df['Fatal (Y/N)']=df['Fatal (Y/N)'].replace('y', 'Y')
df['Fatal (Y/N)']=df['Fatal (Y/N)'].replace('2017', np.NaN)
df['Fatal (Y/N)'] = df['Fatal (Y/N)'].replace('UNKNOWN', 'Unknown')
df['Fatal (Y/N)'] = df['Fatal (Y/N)'].fillna('Unknown')

df.loc[df['Injury'] == 'FATAL', 'Fatal (Y/N)'] = 'Y'
df['Fatal (Y/N)'].value_counts()


We wil keep this column injury as it is, just substituting the Nans with Unknowns

In [ ]:
df['Injury'] = df['Injury'].fillna('Unknown')

## 2.8 Columns Location, Area and Country 

These columns are quite clean, we will just substitute NanNs by "Unknown"

In [ ]:
df['Location'].describe()

In [ ]:
df['Location'].value_counts()

In [ ]:
df['Location'].fillna('unknown', inplace = True)

In [ ]:
df['Area'].value_counts()

In [ ]:
df['Area'].fillna('unknown', inplace = True)

In [ ]:
df['Country'].value_counts()

In [ ]:
df['Country'].fillna('unknown', inplace = True)

In [ ]:
df.info()

## 2.9 Column Type

This column is also almost clean, we just need to do small adjustments

In [ ]:
df['Type'].value_counts()

In [ ]:
df['Type'].fillna('Unknown', inplace = True)

df.loc[df['Type'].str.contains('boat', case=False), 'Type'] = 'Boat'
df.loc[df['Type'].str.contains('Invalid', case=False), 'Type'] = 'Unknown'
df.loc[df['Type'].str.contains('Questionable', case=False), 'Type'] = 'Unknown'

df['Type'].value_counts()

## 2.10 Colums href and href formula

1. We analize the similarity between both columns

In [ ]:
pd.set_option('display.max_rows', 10)  

In [ ]:
df['href formula'].isin(df['href']).value_counts()

2. Explore the differences

In [ ]:
pd.options.display.max_colwidth = 200 # to see all column wide
href_compare = df.loc[df['href formula']!=df['href']]
href_compare[['href formula','href']]

3. Seems that the intention of the column is to have the path http://sharkattackfile.net/spreadsheets/pdf_directory/ plus the name of the pdf, to avoid errors we will substitute the values of both columns by creating new values taking the name of the pdf from the pdf column. Then check if all our values in both colums are identical now


In [ ]:
new_values_href = []
for i in df['pdf']:
    new_values_href.append("http://sharkattackfile.net/spreadsheets/pdf_directory/" + i)

df.drop("href", axis = 1, inplace = True)

df["href"] = new_values_href
df["href formula"] = df["href"]

df['href formula'].isin(df['href']).value_counts()


In [ ]:
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', 10)
df[['href','href formula']]

## 2.11 Column "Investigator or Source"

In [ ]:
pd.set_option('display.max_rows', None)
df['Investigator or Source'].value_counts()

This column seems to have a string containing the information source of the attack, it is not necessary to correct the data, just to substitute NaNs with the word "Unknown"

In [ ]:
df['Investigator or Source'].isna().sum()

In [ ]:
df['Investigator or Source'] = df['Investigator or Source'].fillna('Unknown')

## 2.12 Column Time

In [ ]:
pd.set_option('display.max_rows', None)
df['Time'].value_counts

As this column doesn´t give much relevant info for the analysis of the Dataset we will just replace blank values and non time related values with Nans and then change the Nans to Unknown value

In [ ]:
df['Time'].replace(r'^\s*$', np.NaN, regex=True)
df['Time'].replace('X', np.nan, inplace=True) 
df['Time'].replace('--', np.nan, inplace=True) 
df['Time'] = df['Time'].fillna('Unknown')
df.loc[df['Time'].str.contains('FATAL'), 'Time'] = 'Unknown'

In [ ]:
df['Time'].value_counts()

## 3. Reordering, correction and saving

### 3.1 Reordering

In [ ]:
df.columns

In [ ]:
df = df[['Case Number', 'Case Number.1', 'Date', 'Date Details', 'Year', 'Time', 'Type', 'Country', 'Area',
       'Location', 'Activity', 'Name', 'Sex', 'Age', 'Injury', 'Fatal (Y/N)', 'Species Type', 'Species Info',
         'Investigator or Source','href', 'href formula', 'pdf', 'original order','Unnamed: 23']]

### 3.2 Correction

In [ ]:
def correct_data_cleaning(df):
    
    nota=0 # nota inicial
    
    if len(df.columns)==24:  # si el nº de columnas es 24...
        
        nota+=25
        
        
    if len(df)>=1500:   # si el nº de filas es igual o mayor a 1500...
        
        nota+=25
        
        
    if df.shape==df.dropna().shape:  # si el dataframe es igual al dropna...(no hay nulos)
        
        nota+=25
        
    
    
    # ahora check del tipo de dato.
    
    col_check=0  # variable estado: +1 si todo esta bien (al final debe coincidir con el nº de columnas)
    for c in df:   # para cada columna del dataframe....
        
        col_dtype=df[c].dtype       # type de la columna
        if col_dtype=='object':     # si el type de la columna es 'object', se pone string
            col_dtype=str
        elem_dtype=df[c].map(type)  # aplica el type a la serie, devuelve una serie con el type de cada elemento 
        elem_check=elem_dtype.apply(lambda x: x==col_dtype) # checkea si coincide, devuelve una serie de booleanos, True coincide
        if len(df)==sum(elem_check):  # si el nº de filas coincide con la suma del checkeo, todos los elemento tienen el mismo tipo
            
            col_check+=1
          
    
    if col_check==24:  # si coinciden es que todas las columnas tienen su propio tipo de dato correctamente puesto
        nota+=25
        
        
    return f'Tu nota es {nota}/100'

In [ ]:
correct_data_cleaning(df)

### 3.3 Saving

In [ ]:
df.to_csv('Manuel_Carlos_Cabanillas_Sharks.csv', index=False)